### PR TITLE
Removes empty partitions after dropping rows and splitting datasets

### PR DESCRIPTION
**Pull request by geoffreyangus**
This PR addresses two separate issues: https://github.com/ludwig-ai/ludwig/issues/2324 and https://github.com/ludwig-ai/ludwig/issues/2308.

The issues are addressed by culling empty partitions from the Dask DataFrame at two points: (1) after dropping rows with NaNs (part of the DROP_ROWS missing value strategy) and (2) after splitting the dataset into train/val/test.

In order to maintain/increase performance, we add a persist call at the end of `build_dataset`, which makes it relatively inexpensive to compute the length of partitions repeatedly downstream.

### CODE REVIEW

**Review by arnavgarg1 — APPROVED** [review]
Thanks for making this change - looks good to me!

**Review by ShreyaR — APPROVED** [review]

File: ludwig/data/dataframe/dask.py [hunk 1]
```diff
@@ -129,7 +149,24 @@ def to_ray_dataset(self, df):
         return from_dask(df)
 
     def from_ray_dataset(self, dataset) -> dd.DataFrame:
-        return dataset.to_dask()
+        """Custom Ray to Dask conversion implementation to pass in meta during dd.DataFrame creation."""
```
Comment (arnavgarg1) [inline ludwig/data/dataframe/dask.py:152]: @geoffreyangus should we revert back to `dataset.to_dask()` if the empty partitions issue doesn't stem from a need to pass in meta?

Comment (geoffreyangus) [reply ludwig/data/dataframe/dask.py]: Yup, reverted!

**Comment by arnavgarg1** [comment]
LGTM!

**Pull request merged**